src/v/cloud_topics/frontend/frontend.cc

```diff
@@ -468,24 +468,35 @@ ss::future<> bg_upload_and_replicate(
                 if (res.has_error()) {
                     return res.error();
                 }
-                // We know that the data is replicated so it's safe to add
-                // the batch to the record batch cache before returning.
                 if (cache_enabled) {
-                    vassert(
-                      res.value().last_term != model::term_id{},
-                      "Term not set");
-                    update_batches(
-                      inp,
-                      kafka::offset_cast(res.value().last_offset),
-                      res.value().last_term);
-                    for (const auto& b : inp) {
+                    // The term_id is not guaranteed to be set if there was an
+                    // idempotency violation and the batch was deduplicated. In
+                    // this case the finished request is picked up by the
+                    // rm_stm. But in this case there is a chance that the batch
```

Comment:
Is this accurate? If there was an idempotency violation I'd expect result.has_error() = true and L469 to be taken?

Atleast from the API I feel like in all cases where result.has_error() = false, a valid term should be set.

Reply:
I went through the whole comment again (and the Slack thread). AFAICT, there's no idempotency violation here. There appears to be a request replay (the same request was retried multiple times by the client). Between retries, the broker likely restarted, and the request was served from the snapshot of completed requests taken before the reboot of the partition. As I mentioned in the thread, the cleanest solution would be to include terms in the snapshot for finished requests. That said, this might be acceptable as-is, since the cache is primarily a performance optimization.

Reply:
Maybe a good discussion topic for CT meeting tomorrow?

Reply:
> the same request was retried multiple times by the client
This is what I saw, maybe I was incorrect to call it idempotency violation. What happened is that the request from the finished request list was used and the resulting record batch that we got in the happy path didn't have term set.
I think it is acceptable to just ignore the batch because it is an optimization.